src/gui/plugins/copy_paste/CopyPaste.hh

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef IGNITION_GAZEBO_GUI_COPYPASTE_HH_
+#define IGNITION_GAZEBO_GUI_COPYPASTE_HH_
+
+#include <memory>
+
+#include <ignition/msgs.hh>
+
+#include "ignition/gazebo/gui/GuiSystem.hh"
+
+namespace ignition
+{
+namespace gazebo
+{
+  class CopyPastePrivate;
+
+  /// \brief Plugin for copying/pasting entities in the GUI.
+  class CopyPaste : public ignition::gazebo::GuiSystem
+  {
+    Q_OBJECT
+
+    /// \brief Constructor
+    public: CopyPaste();
+
+    /// \brief Destructor
+    public: ~CopyPaste() override;
+
+    // Documentation inherited
+    public: void LoadConfig(const tinyxml2::XMLElement *_pluginElem) override;
+
+    // Documentation inherited
+    public: void Update(const UpdateInfo &_info,
+                EntityComponentManager &_ecm) override;
+
+    /// \brief Callback to copy the selected entity
+    public slots: void OnCopy();
+
+    /// \brief Callback to paste the data that has been copied, if copied data
+    /// exists.
+    public slots: void OnPaste();
+
+    // Documentation inherited
+    protected: bool eventFilter(QObject *_obj, QEvent *_event) override;
+
+    /// \brief Callback for handling a copy service request
+    /// \param[in] _req The service request, which contains the name of the
+    /// entity to be copied
+    /// \param[out] _resp The service response
+    private: bool CopyServiceCB(const ignition::msgs::StringMsg &_req,
+                ignition::msgs::Boolean &_resp);
+
+    /// \brief Callback for handling a paste service request
+    /// \param[in] _req The service request
+    /// \param[out] _resp The service response
+    private: bool PasteServiceCB(const ignition::msgs::Empty &_req,
+                ignition::msgs::Boolean &_resp);
+
+    /// \internal
+    /// \brief Pointer to private data
+    private: std::unique_ptr<CopyPastePrivate> dataPtr;
```

Comment:
`IGN_UTILS_IMPL_PTR` ?

Reply:
I tried using it, but I don't think it will work since some of the things I am using (`transport::node`, `std::mutex`) don't appear to be copyable by default. But, maybe I don't understand `IGN_UTILS_IMPL_PTR` enough. Here's the error output I saw when I tried using `IGN_UTILS_IMPL_PTR`:

<details>
  <summary>Compilation errors</summary>

```
Starting >>> ignition-gazebo6
--- stderr: ignition-gazebo6                             
In file included from /usr/include/ignition/utils1/ignition/utils/ImplPtr.hh:24,
                 from /ws/install/include/ignition/gui6/ignition/gui/GuiEvents.hh:30,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:24:
/usr/include/ignition/utils1/ignition/utils/detail/DefaultOps.hh: In instantiation of 'T* ignition::utils::detail::DefaultCopyConstruct(const T&) [with T = ignition::gazebo::CopyPaste::Implementation]':
/usr/include/ignition/utils1/ignition/utils/detail/ImplPtr.hh:120:21:   required from 'ignition::utils::ImplPtr<T> ignition::utils::MakeImpl(Args&& ...) [with T = ignition::gazebo::CopyPaste::Implementation; Args = {}]'
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:58:58:   required from here
/usr/include/ignition/utils1/ignition/utils/detail/DefaultOps.hh:56:16: error: use of deleted function 'ignition::gazebo::CopyPaste::Implementation::Implementation(const ignition::gazebo::CopyPaste::Implementation&)'
         return new T(_source);
                ^~~~~~~~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: note: 'ignition::gazebo::CopyPaste::Implementation::Implementation(const ignition::gazebo::CopyPaste::Implementation&)' is implicitly deleted because the default definition would be ill-formed:
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: use of deleted function 'ignition::transport::v11::Node::Node(const ignition::transport::v11::Node&)'
In file included from /usr/include/ignition/transport11/ignition/transport.hh:31,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:27:
/usr/include/ignition/transport11/ignition/transport/Node.hh:95:38: note: 'ignition::transport::v11::Node::Node(const ignition::transport::v11::Node&)' is implicitly deleted because the default definition would be ill-formed:
     class IGNITION_TRANSPORT_VISIBLE Node
                                      ^~~~
/usr/include/ignition/transport11/ignition/transport/Node.hh:95:38: error: use of deleted function 'std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = ignition::transport::v11::NodePrivate; _Dp = std::default_delete<ignition::transport::v11::NodePrivate>]'
In file included from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/unique_ptr.h:397:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
       ^~~~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: use of deleted function 'std::mutex::mutex(const std::mutex&)'
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/mutex:43,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:19:
/usr/include/c++/8/bits/std_mutex.h:97:5: note: declared here
     mutex(const mutex&) = delete;
     ^~~~~
In file included from /usr/include/ignition/utils1/ignition/utils/ImplPtr.hh:24,
                 from /ws/install/include/ignition/gui6/ignition/gui/GuiEvents.hh:30,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:24:
/usr/include/ignition/utils1/ignition/utils/detail/DefaultOps.hh: In instantiation of 'void ignition::utils::detail::DefaultCopyAssign(T&, const T&) [with T = ignition::gazebo::CopyPaste::Implementation]':
/usr/include/ignition/utils1/ignition/utils/detail/ImplPtr.hh:120:21:   required from 'ignition::utils::ImplPtr<T> ignition::utils::MakeImpl(Args&& ...) [with T = ignition::gazebo::CopyPaste::Implementation; Args = {}]'
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:58:58:   required from here
/usr/include/ignition/utils1/ignition/utils/detail/DefaultOps.hh:64:15: error: use of deleted function 'ignition::gazebo::CopyPaste::Implementation& ignition::gazebo::CopyPaste::Implementation::operator=(const ignition::gazebo::CopyPaste::Implementation&)'
         _dest = _source;
         ~~~~~~^~~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: note: 'ignition::gazebo::CopyPaste::Implementation& ignition::gazebo::CopyPaste::Implementation::operator=(const ignition::gazebo::CopyPaste::Implementation&)' is implicitly deleted because the default definition would be ill-formed:
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: use of deleted function 'ignition::transport::v11::Node& ignition::transport::v11::Node::operator=(const ignition::transport::v11::Node&)'
In file included from /usr/include/ignition/transport11/ignition/transport.hh:31,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:27:
/usr/include/ignition/transport11/ignition/transport/Node.hh:95:38: note: 'ignition::transport::v11::Node& ignition::transport::v11::Node::operator=(const ignition::transport::v11::Node&)' is implicitly deleted because the default definition would be ill-formed:
     class IGNITION_TRANSPORT_VISIBLE Node
                                      ^~~~
/usr/include/ignition/transport11/ignition/transport/Node.hh:95:38: error: use of deleted function 'std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = ignition::transport::v11::NodePrivate; _Dp = std::default_delete<ignition::transport::v11::NodePrivate>]'
In file included from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/unique_ptr.h:398:19: note: declared here
       unique_ptr& operator=(const unique_ptr&) = delete;
                   ^~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: no matching function for call to 'std::__cxx11::basic_string<char>::operator=(const string&) const'
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/string:52,
                 from /usr/include/c++/8/stdexcept:39,
                 from /usr/include/c++/8/array:39,
                 from /usr/include/c++/8/tuple:39,
                 from /usr/include/c++/8/bits/unique_ptr.h:37,
                 from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/basic_string.h:664:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' <near match>
       operator=(const basic_string& __str)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:664:7: note:   passing 'const string*' {aka 'const std::__cxx11::basic_string<char>*'} as 'this' argument discards qualifiers
/usr/include/c++/8/bits/basic_string.h:732:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' <near match>
       operator=(basic_string&& __str)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:732:7: note:   conversion of argument 1 would be ill-formed:
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: cannot bind rvalue reference of type 'std::__cxx11::basic_string<char>&&' to lvalue of type 'const string' {aka 'const std::__cxx11::basic_string<char>'}
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/string:52,
                 from /usr/include/c++/8/stdexcept:39,
                 from /usr/include/c++/8/array:39,
                 from /usr/include/c++/8/tuple:39,
                 from /usr/include/c++/8/bits/unique_ptr.h:37,
                 from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/basic_string.h:703:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const _CharT*) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(const _CharT* __s)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:714:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(_CharT) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(_CharT __c)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:795:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::initializer_list<_Tp>) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(initializer_list<_CharT> __l)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:809:8: note: candidate: 'template<class _Tp> std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_If_sv<_Tp, std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&> std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const _Tp&) [with _Tp = _Tp; _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
        operator=(const _Tp& __svt)
        ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:809:8: note:   template argument deduction/substitution failed:
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: no matching function for call to 'std::__cxx11::basic_string<char>::operator=(const string&) const'
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/string:52,
                 from /usr/include/c++/8/stdexcept:39,
                 from /usr/include/c++/8/array:39,
                 from /usr/include/c++/8/tuple:39,
                 from /usr/include/c++/8/bits/unique_ptr.h:37,
                 from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/basic_string.h:664:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' <near match>
       operator=(const basic_string& __str)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:664:7: note:   passing 'const string*' {aka 'const std::__cxx11::basic_string<char>*'} as 'this' argument discards qualifiers
/usr/include/c++/8/bits/basic_string.h:732:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' <near match>
       operator=(basic_string&& __str)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:732:7: note:   conversion of argument 1 would be ill-formed:
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: cannot bind rvalue reference of type 'std::__cxx11::basic_string<char>&&' to lvalue of type 'const string' {aka 'const std::__cxx11::basic_string<char>'}
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/string:52,
                 from /usr/include/c++/8/stdexcept:39,
                 from /usr/include/c++/8/array:39,
                 from /usr/include/c++/8/tuple:39,
                 from /usr/include/c++/8/bits/unique_ptr.h:37,
                 from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/basic_string.h:703:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const _CharT*) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(const _CharT* __s)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:714:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(_CharT) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(_CharT __c)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:795:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::initializer_list<_Tp>) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(initializer_list<_CharT> __l)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:809:8: note: candidate: 'template<class _Tp> std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_If_sv<_Tp, std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&> std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const _Tp&) [with _Tp = _Tp; _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
        operator=(const _Tp& __svt)
        ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:809:8: note:   template argument deduction/substitution failed:
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: use of deleted function 'std::mutex& std::mutex::operator=(const std::mutex&)'
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/mutex:43,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:19:
/usr/include/c++/8/bits/std_mutex.h:98:12: note: declared here
     mutex& operator=(const mutex&) = delete;
            ^~~~~~~~
make[2]: *** [src/gui/plugins/copy_paste/CMakeFiles/CopyPaste.dir/CopyPaste.cc.o] Error 1
make[1]: *** [src/gui/plugins/copy_paste/CMakeFiles/CopyPaste.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
---
Failed   <<< ignition-gazebo6 [8.28s, exited with code 2]

Summary: 0 packages finished [8.58s]
  1 package failed: ignition-gazebo6
  1 package had stderr output: ignition-gazebo6
```

</details>